JBPM-7267: Improve BB computation performance

romartin opened the pull request:
Hey @hasys 
This is the new version for this https://github.com/kiegroup/lienzo-core/pull/35, which includes the tests. Tests are here https://github.com/kiegroup/lienzo-tests/pull/25
Thanks!

hasys commented on src/main/java/com/ait/lienzo/client/core/types/BoundingPoints.java:
```diff
@@ -21,6 +21,8 @@
 import java.util.Collections;
 import java.util.Iterator;
 
+import com.ait.tooling.nativetools.client.util.Console;
```
This new import statement is not in use.

romartin commented on src/main/java/com/ait/lienzo/client/core/types/BoundingPoints.java:
```diff
@@ -35,13 +37,40 @@ public BoundingPoints(final BoundingBox bbox)
+        double x1;
+        double y1;
+
+        if (bbox.getMaxX() > bbox.getMinX())
```
well @hasys I think the current approach is fine... I mean the BoundingBox class handles just 4 points (coordinates) internally, and the intersect method uses these absolute coordinates, so that seems fine, it does not understand about direction, there is no need for... i'ts just an state object.
On the other hand the BoundingPoints class takes care of the direction, and its state is an array of points. So it depends on the way you need to obtain the points and the goal you want to achieve... but IMO is fine having as two separate classes.

romartin commented on src/main/java/com/ait/lienzo/client/core/shape/BezierCurve.java:
```diff
@@ -77,7 +77,23 @@ protected BezierCurve(final JSONObject node, final ValidationContext ctx) throws
     public BoundingBox getBoundingBox()
     {
-        final BoundingBox bbox = Geometry.getBoundingBox(this);
+        Point2DArray p = getPoint2DArray();
```
well I don't think so...  the point array is just the input to calculate the BB, given by the result of the call to Geometry.getBoundingBoxOfCubicCurve, but it does not mean that the BB points will be the first and last one from the array... isn't it?

romartin commented on src/main/java/com/ait/lienzo/client/core/types/BoundingPoints.java:
```diff
@@ -35,13 +37,40 @@ public BoundingPoints(final BoundingBox bbox)
+
+        if (bbox.getMaxX() > bbox.getMinX())
+        {
+            x0 = x + 0;
```
The BoundingBox does not understand about directions, IMO there is no need for, the box and the points resulting are two different things, because as you say, direction matters, so the points for the box depend on it....

romartin commented:
Hey @hasys ,

I have been trying to answer your Qs, hope that makes sense for you. Also answered in https://github.com/kiegroup/lienzo-tests/pull/25 

Just lemme know your feedback plz!

Thx!

hasys commented:
Hi @romartin,

probably I wasn't clear.

Let me explain it with code :) According to this if statement: https://github.com/kiegroup/lienzo-core/pull/40/files/70c2516aa5b9d36afaa9019a2604bef0b3f14502#diff-fe6ecfa329e875cd2500dcd0c71b5b15R45 it is possible that Min point's coordinate will be bigger then Max in reality. But if it is therefore this method https://github.com/kiegroup/lienzo-core/pull/40/files#diff-c4699f925c6f0f21fe6f6c9e27802ed6R223 will fail in some cases.

I wrote `testIntersection` test to check my assumption and it looks like my assumption was correct. My test is failing on third assert, but it shouldn't:
```java
@RunWith(LienzoMockitoTestRunner.class)
public class BoundingBoxTest
{
    [...]

    @Test
    public void testIntersect()
    {
        BoundingBox box1 = new BoundingBox(0, 0, 100, 100);
        BoundingBox box2 = new BoundingBox(50, 50, 150, 150);

        assertTrue(box1.intersects(box2));
        assertTrue(box2.intersects(box1));

        box1 = new BoundingBox(100, 100, 0, 0);
        assertTrue(box1.intersects(box2));
    }
}
```

So, my suggestion to move logic about "reversed" min and max points to BoundingBox get/set methods and constructor will protect us from if/else statements for external uses and will protect internal BoundingBox staff itself from some issues as well.

Does it make sense or I am missing something?

romartin commented:
Sorry closed this issue by error (I removed accidentally my branch for it) , re-opening!

romartin commented:
Hey @hasys 

Good catches about the `BoundingBox`/`BoundingPoints` stuff - Added a new commit which fixes this, tests and also rebased, hope it makes sense for you now. 

The idea, as you can see for example [here](https://github.com/kiegroup/lienzo-core/blob/master/src/main/java/com/ait/lienzo/client/core/types/BoundingBox.java#L351), is that the BB class already handles the min/max internal fields and guarantees they're in the right order. 

The issue appears when creating an instance by using the constructor that takes 4 `double` parameters, as it's just creating the overlay object and do not perform the min/max evaluations (not calling `addX/Y` methods), so fixed it. Also after this, no need on the `BoundingPoints` for doing min/max checks anymore, so reverted that code as well, good catches!

If I'm fine, at this point, we're only missing the test cases for some methods on `Geometry` class, right?  https://github.com/kiegroup/lienzo-tests/pull/25 Already working on those!!

Thanks!!

romartin commented:
Jenkins please retest this

romartin commented:
@hasys  on the other hand, we should try to merge this ASAP please. For two reasons: first because it's a performance improvement that has really an impact on some browsers (FF/IE), and also because this is already merged on `7.7.x` and working fine, so we need to merge this to have both branches on sync...
I fixed the BB issues related to "negative" locations and also added tests for all method affected in `Geometry`.
Thanks in advance!

hasys commented on src/main/java/com/ait/lienzo/client/core/util/Geometry.java:
```diff
@@ -1403,8 +1377,46 @@ public static void getCardinalIntersects(PathPartList path, Point2DArray cardina
                     segmentStart = end;
+                    break;
+                }
+                case PathPartEntryJSO.BEZIER_CURVETO_ABSOLUTE:
```
Some branches of this Switch is tested but this Case didn't.

romartin replied:
good catch! done - as now I'm testing a success intersection between curves and line, so it's being called :+1:

hasys commented on src/main/java/com/ait/lienzo/client/core/util/Geometry.java:
```diff
@@ -371,194 +228,319 @@ public static BoundingBox getBoundingBoxOfArc(final Point2D ps, final Point2D pc
-            final double r = Math.round(findRoots(derivative, t, values) / NRRF_PRECISION) * NRRF_PRECISION;
-
-            if (root.contains(r))
+            if (greaterOrCloseEnough(t, 0) && lesserOrCloseEnough(t, 1) && greaterOrCloseEnough(s, 0) && lesserOrCloseEnough(s, 1))
```
I guess we can add some method like:
```java
public void static bool isBetween(double min, double value, double max)
{
    return greaterOrCloseEnough(value, min) && lesserOrCloseEnough(value, max);
}
```

romartin commented on src/main/java/com/ait/lienzo/client/core/shape/BezierCurve.java:
```diff
@@ -77,7 +77,23 @@ protected BezierCurve(final JSONObject node, final ValidationContext ctx) throws
     @Override
     public BoundingBox getBoundingBox()
     {
-        final BoundingBox bbox = Geometry.getBoundingBox(this);
```
done

romartin commented:
@hasys thanks for the review!! Applied your comments again and added the missing test cases. I also rebased to check everything is fine with latest upstream.
Thanks!

tiagodolphine commented:
hey @romartin @hasys what is missing on this PR to be merged? 
I think it would be great to merge this before your PTO @romartin this is necessary for the repos sync kie and ahome.

hasys commented:
Hi @romartin, @tiagodolphine,

looks good to me now. Thank you @romartin, great job!
